sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -315,26 +333,200 @@ public virtual PartitionProperties CreatePartitionPropertiesFromResponse(AmqpMes
         private static AmqpMessage BuildAmqpMessageFromEvent(EventData source,
                                                              string partitionKey)
         {
-            if (!MemoryMarshal.TryGetArray(source.EventBody.ToMemory(), out var bodySegment))
+            var sourceMessage = source.GetRawAmqpMessage();
+
+            var message = sourceMessage switch
             {
-                bodySegment = new ArraySegment<byte>(source.EventBody.ToArray());
+                _ when sourceMessage.Body.TryGetData(out var dataBody) => AmqpMessage.Create(TranslateDataBody(dataBody)),
+                _ when sourceMessage.Body.TryGetSequence(out var sequenceBody) => AmqpMessage.Create(TranslateSequenceBody(sequenceBody)),
+                _ when sourceMessage.Body.TryGetValue(out var valueBody) => AmqpMessage.Create(TranslateValueBody(valueBody)),
+                _ => AmqpMessage.Create(new Data { Value = new ArraySegment<byte>(Array.Empty<byte>()) })
```

Comment:
For cases where the body type cannot be determined, assume an empty data body;  this is necessary to preserve consistency with existing behavior.